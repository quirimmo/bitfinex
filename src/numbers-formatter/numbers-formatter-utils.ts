export const getDefaultPropsValue = (
	defaultShowNaN: boolean,
	defaultDecimalPrecision: number,
	showNaN?: boolean,
	decimalPrecision?: number
): { showNaN: boolean; decimalPrecision: number } => {
	// get the showNaN value if provided, otherwise set default to true
	const showNaNValue: boolean = typeof showNaN === 'undefined' ? defaultShowNaN : Boolean(showNaN);
	// get the privded decimal precision or a default one
	const decimalPrecisionValue: number = decimalPrecision || defaultDecimalPrecision;
	return { showNaN: showNaNValue, decimalPrecision: decimalPrecisionValue };
};

export const getIntegerAndDecimals = (numericValue: number, decimalPrecision: number): { integerValues: string; decimalValues: string } => {
	// convert the number to a string, fixing the decimal precision
	const stringValue: string = numericValue.toFixed(decimalPrecision);
	// get the integer part of the number
	const integerValues: string = stringValue.split('.')[0];
	// get the decimal part of the number if any, otherwise an empty string
	const decimalValues: string = stringValue.split('.')[1] || '';
	return { integerValues, decimalValues };
};

export const composeFormattedValue = (formattedIntegers: string[], decimalValues: string): string => {
	// get back a string from the array of integers
	let ret = formattedIntegers.join('');
	// if there are decimal values, suffix them at the end after a dot
	if (decimalValues.length) {
		ret += `.${decimalValues}`;
	}
	return ret;
};

export const getFormattedIntegers = (integerValues: string): string[] => {
	// reduce the array from right to left, appending commas, and return the built array
	return integerValues.split('').reduceRight(onReduceRight, []);

	function onReduceRight(accumulator: string[], currentValue: string, index: number) {
		// use the idea of a queue to enqueue the element always at the top of the array,
		// for keeeping the numeric value of the starting number
		accumulator.unshift(currentValue);
		// enque the comma when thousand is reached and the number is not finished (e.g. avoid ,100,000)
		if (countNumbersExceptCommas(accumulator) % 3 === 0 && index > 0) {
			accumulator.unshift(',');
		}
		return accumulator;
	}

	function countNumbersExceptCommas(input: string[]): number {
		return input.filter((el: string) => el !== ',').length;
	}
};
