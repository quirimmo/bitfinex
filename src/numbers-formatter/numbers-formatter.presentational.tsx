import * as React from 'react';
import { getDefaultPropsValue, getIntegerAndDecimals, composeFormattedValue, getFormattedIntegers } from './numbers-formatter-utils';

export interface INumbersFormatterProps {
	value: number | string;
	decimalPrecision?: number;
	showNaN?: boolean;
	children?(formattedValue: string): JSX.Element;
}

export interface INumbersFormatterState {
	formattedValue: string;
}

export class NumbersFormatter extends React.Component<INumbersFormatterProps, INumbersFormatterState> {
	public static readonly DEFAULT_DECIMAL_PRECISION: number = 2;
	public static readonly DEFAULT_SHOW_NAN: boolean = true;

	constructor(props: INumbersFormatterProps) {
		super(props);
		this.state = { formattedValue: '' };
		this.formatValue = this.formatValue.bind(this);
	}

	public componentDidMount() {
		const { showNaN, decimalPrecision } = getDefaultPropsValue(
			NumbersFormatter.DEFAULT_SHOW_NAN,
			NumbersFormatter.DEFAULT_DECIMAL_PRECISION,
			this.props.showNaN,
			this.props.decimalPrecision
		);
		this.formatValue(showNaN, decimalPrecision);
	}

	public formatValue(showNaN: boolean, decimalPrecision: number): void {
		// get the number provided as value
		const numericValue = Number(this.props.value);
		// if it is not a number
		if (isNaN(numericValue)) {
			console.error('Error with the provided number:', this.props.value);
			// if showNaN true, display NaN. Otherwise keep empty
			const formattedValue: string = showNaN ? NaN.toString() : '';
			// refresh the state with the value to display
			this.setState({ formattedValue });
		} else {
			const { integerValues, decimalValues } = getIntegerAndDecimals(numericValue, decimalPrecision);
			// create an array of all the integer values in order, appending the commas
			const formattedIntegers: string[] = getFormattedIntegers(integerValues);
			// compose the final string
			const ret: string = composeFormattedValue(formattedIntegers, decimalValues);
			// refresh the state with the formatted value to display
			this.setState({ formattedValue: ret });
		}
	}

	public render() {
		return (
			<React.Fragment>
				{this.props.children && this.props.children(this.state.formattedValue)}
				{!this.props.children && this.state.formattedValue}
			</React.Fragment>
		);
	}
}
