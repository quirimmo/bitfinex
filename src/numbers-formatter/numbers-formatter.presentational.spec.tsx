import * as React from 'react';
import { shallow } from 'enzyme';

const mockGetDefaultPropsValue: any = jest.fn().mockReturnValue({ showNaN: true, decimalPrecision: 2 });
const mockGetIntegerAndDecimals: any = jest.fn().mockReturnValue({ integerValues: '4324321121', decimalValues: '433' });
const mockComposeFormattedValue: any = jest.fn().mockReturnValue('4,324,321,121.43');
const mockGetFormattedIntegers: any = jest.fn().mockReturnValue(['4', ',', '324', '321', '121']);
jest.mock('./numbers-formatter-utils', () => ({
	getDefaultPropsValue: mockGetDefaultPropsValue,
	getIntegerAndDecimals: mockGetIntegerAndDecimals,
	composeFormattedValue: mockComposeFormattedValue,
	getFormattedIntegers: mockGetFormattedIntegers
}));

import { NumbersFormatter } from './numbers-formatter.presentational';

let component: any;
let instance: any;
let spyFormatValue: any = jest.fn();

describe('NumbersFormatter Presentational Component', () => {
	beforeEach(() => {
		jest.useFakeTimers();
		spyFormatValue = jest.spyOn(NumbersFormatter.prototype, 'formatValue');
		component = shallow(
			<NumbersFormatter value={4324321121.433}>{(formattedValue: string) => <div>I am the formatted value: {formattedValue}</div>}</NumbersFormatter>
		);
		instance = component.instance();
	});
	afterEach(() => {
		jest.runOnlyPendingTimers();
		jest.clearAllMocks();
	});

	it('should be defined', () => {
		expect(component).toBeDefined();
	});

	it('should match the snapshot', () => {
		expect(component).toMatchSnapshot();
	});

	it('should define the instance methods', () => {
		expect(typeof instance.formatValue).toEqual('function');
	});

	it('should define the state formattedValue prop', () => {
		expect(component.state().formattedValue).toBeDefined();
	});

	it('should display the final text', () => {
		expect(component.text()).toEqual('I am the formatted value: 4,324,321,121.43');
	});

	describe('componentDidMount', () => {
		it('should call the getDefaultPropsValue function', () => {
			expect(mockGetDefaultPropsValue).toHaveBeenCalledWith(true, 2, undefined, undefined);
		});

		it('should call the formatValue method', () => {
			expect(spyFormatValue).toHaveBeenCalledWith(true, 2);
		});
	});

	describe('formatValue', () => {
		describe('not numberic value prop', () => {
			let spy: jest.SpyInstance;
			beforeEach(() => {
				spy = jest.spyOn(console, 'error').mockImplementation(() => {});
			});
			afterEach(() => {
				spy.mockRestore();
			});

			it('should log the error if the value is not a valid number', () => {
				component.setProps({ value: 'test' });
				instance.componentDidMount();
				expect(spy).toHaveBeenCalledWith('Error with the provided number:', 'test');
			});

			it('should set the formattedValue in the state showing NaN', () => {
				component.setProps({ value: 'test', showNaN: true });
				const spySetState: any = jest.spyOn(instance, 'setState');
				instance.componentDidMount();
				expect(spySetState).toHaveBeenCalledWith({ formattedValue: 'NaN' });
			});

			it('should set the formattedValue in the state without showing NaN', () => {
				component.setProps({ value: 'test', showNaN: false });
				mockGetDefaultPropsValue.mockReturnValue(() => ({ showNaN: false, decimalPrecision: 2 }));
				const spySetState: any = jest.spyOn(instance, 'setState');
				instance.componentDidMount();
				expect(spySetState).toHaveBeenCalledWith({ formattedValue: '' });
			});
		});

		describe('numberic value prop', () => {
			it('should call the getIntegerAndDecimals function', () => {
				instance.formatValue();
				expect(mockGetIntegerAndDecimals).toHaveBeenCalled();
			});

			it('should call the getFormattedIntegers function', () => {
				instance.formatValue();
				expect(mockGetFormattedIntegers).toHaveBeenCalled();
			});

			it('should call the composeFormattedValue function', () => {
				instance.formatValue();
				expect(mockComposeFormattedValue).toHaveBeenCalled();
			});
		});
	});
});
