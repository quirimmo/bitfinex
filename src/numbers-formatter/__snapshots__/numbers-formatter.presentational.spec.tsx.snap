// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NumbersFormatter Presentational Component should match the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <NumbersFormatter
    value={4324321121.433}
  >
    [Function]
  </NumbersFormatter>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <div>
          I am the formatted value: 
          4,324,321,121.43
        </div>,
        false,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "I am the formatted value: ",
            "4,324,321,121.43",
          ],
        },
        "ref": null,
        "rendered": Array [
          "I am the formatted value: ",
          "4,324,321,121.43",
        ],
        "type": "div",
      },
      false,
    ],
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <div>
            I am the formatted value: 
            4,324,321,121.43
          </div>,
          false,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "I am the formatted value: ",
              "4,324,321,121.43",
            ],
          },
          "ref": null,
          "rendered": Array [
            "I am the formatted value: ",
            "4,324,321,121.43",
          ],
          "type": "div",
        },
        false,
      ],
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
